tests/testthat/test-root.R

```diff
@@ -156,20 +156,50 @@ test_that("is_svn_root", {
   )
 })
 
-test_that("is_git_root", {
+setup_git_root <- function() {
   temp_dir <- tempfile("git")
   unzip("vcs/git.zip", exdir = temp_dir)
-  wd <- normalizePath(temp_dir, winslash = "/")
+  return(temp_dir)
+}
 
+test_that("is_git_root", {
+  temp_dir <- setup_git_root()
+  wd <- normalizePath(temp_dir, winslash = "/")
   hierarchy <- function(n = 0L) {
     do.call(file.path, list(wd, "git", "a", "b", "c")[seq_len(n + 1L)])
   }
-
+  path <- hierarchy(4L)
   stop_path <- normalizePath(tempdir(), winslash = "/")
+
+  mockr::with_mock(
+    is_root = function(x) x == stop_path,
+    expect_equal(find_root(is_git_root, path = path), hierarchy(1L)),
+    expect_equal(find_root(is_vcs_root, path = path), hierarchy(1L)),
+    expect_error(find_root(is_git_root, path = hierarchy(0L)),
+                 "No root directory found.* a directory `.*`"),
+    expect_error(find_root(is_vcs_root, path = hierarchy(0L)),
+                 "No root directory found.* a directory `.*`"),
+    TRUE
+  )
+})
+
+test_that("is_git_root for separated git directory", {
+  temp_dir <- setup_git_root()
+  wd <- normalizePath(temp_dir, winslash = "/")
+  hierarchy <- function(n = 0L) {
+    do.call(file.path, list(wd, "git", "a", "b", "c")[seq_len(n + 1L)])
+  }
   path <- hierarchy(4L)
+  stop_path <- normalizePath(tempdir(), winslash = "/")
 
-  with_mock(
-    `rprojroot:::is_root` = function(x) x == stop_path,
+  # Copy .git dir to a separate location, then make a .git file.
+  # (other_git_folder becomes a bare git repo)
+  old_git_location <- file.path(wd, "git", ".git")
+  new_git_location <- file.path(wd, "other_git_folder")
+  file.rename(old_git_location, new_git_location)
+  writeLines(paste("gitdir:", new_git_location), old_git_location)
```

Comment:
Maybe add this setup code as an option to `setup_git_root()`? If the tests are identical in the two cases, we could also extract a function.